### ISSUE TITLE
limit bandwidth option

### ISSUE THREAD

**Issue by gregf** [restic/restic#1216]
It would be nice if restic offered an option like rsync has to limit bandwidth usage. In particular when uploading. I have pretty bad internet where I live and when I upload it really kills my download connection. It would be great if I could limit the upload bandwidth being used by restic.

**Comment by fd0** [comment]
Great idea, I'll keep this issue as a reminder.

**Comment by flo82** [comment]
i'm using traffic shaping with trafficcontrol/iptables.
have a look at this link: https://help.atmail.com/hc/en-us/articles/201566464-Throttling-Bandwidth-using-Traffic-Controller-for-Linux

works perfect :-)

**Comment by nibble** [comment]
An integrated bandwith limiting feature would be great.

In the meantime, what I use to limit traffic for the sftp backend is to pipe the connection through `pv`:

```
Host myserver-restic
        ProxyCommand pv -q -L 100k | nc myserver.com 22
```

**Comment by apirogov** [comment]
I would also really like to have this. My upload speed is about 120KB/s and for an initial backup of 300GB to BackBlaze I would need a month at full upload speed, so I'd rather make two months at half speed out of it and use my internet normally.

Off-topic - is restic even designed for what I want to do? Can the upload be interrupted and continued? I can not assume to have my laptop powered and online for such a long time.

**Comment by fd0** [comment]
> Off-topic - is restic even designed for what I want to do? Can the upload be interrupted and continued? I can not assume to have my laptop powered and online for such a long time.

While it's not an ideal situation, I'm confident that restic will handle this just fine. It just might take longer to upload the data, and it will upload duplicate data.

During upload, even if the backup is not finished, restic will write intermediate index files that contain a list of blobs uploaded so far. At a minimum, such an index file is written every five minutes.

When you abort the backup, and re-run it later, restic will first download all the index files so it knows which blobs are already in the repo. This includes the intermediate index files.

So, eventually you'll finish uploading all data.

**Comment by armhold** [comment]
This would be a really nice feature to have. 

> While it's not an ideal situation, I'm confident that restic will handle this just fine. It just might take longer to upload the data, and it will upload duplicate data.

This is in fact how I uploaded my 185GB to B2. Since it took several days of saturating my Internet, I would ^C the process during the day, and re-run the backup at night. Not ideal, but it worked OK after a few days and then some prune/check sessions.

For pruning, I've found that merely suspending the process (^Z) and then "fg" works fine, even after several hours of suspension.

**Comment by fd0** [comment]
> For pruning, I've found that merely suspending the process (^Z) and then "fg" works fine, even after several hours of suspension.

Please run `restic check --read-data` on that repo, I'm not sure what `prune` does when you suspend it for so long...

**Comment by armhold** [comment]
> Please run restic check --read-data on that repo (re-downloads and checks everything), I'm not sure what prune does when you suspend it for so long...

Thanks for the heads-up. It checked out fine.

**Comment by fd0** [comment]
I've just merged #1336, which adds the options `--limit-upload` and `--limit-download` :)
